rai/qt/qt.cpp

```diff
@@ -717,6 +720,91 @@ wallet (wallet_a)
 	});
 }
 
+rai_qt::stats_viewer::stats_viewer (rai_qt::wallet & wallet_a) :
+window (new QWidget),
+layout (new QVBoxLayout),
+model (new QStandardItemModel),
+view (new QTableView),
+refresh (new QPushButton ("Refresh")),
+back (new QPushButton ("Back")),
+wallet (wallet_a)
+{
+	model->setHorizontalHeaderItem (0, new QStandardItem ("Last updated"));
+	model->setHorizontalHeaderItem (1, new QStandardItem ("Type"));
+	model->setHorizontalHeaderItem (2, new QStandardItem ("Detail"));
+	model->setHorizontalHeaderItem (3, new QStandardItem ("Direction"));
+	model->setHorizontalHeaderItem (4, new QStandardItem ("Value"));
+	view->setModel (model);
+	view->setEditTriggers (QAbstractItemView::NoEditTriggers);
+	view->verticalHeader ()->hide ();
+	view->horizontalHeader ()->setStretchLastSection (true);
+	layout->setContentsMargins (0, 0, 0, 0);
+	layout->addWidget (view);
+	layout->addWidget (refresh);
+	layout->addWidget (back);
+	window->setLayout (layout);
+
+	QObject::connect (back, &QPushButton::released, [this]() {
+		this->wallet.pop_main_stack ();
+	});
+	QObject::connect (refresh, &QPushButton::released, [this]() {
+		refresh_stats ();
+	});
+
+	refresh_stats ();
+}
+
+void rai_qt::stats_viewer::refresh_stats ()
+{
+	model->removeRows (0, model->rowCount ());
+
+	auto sink = wallet.node.stats.log_sink_json ();
+	wallet.node.stats.log_counters (*sink);
+	auto json = static_cast<boost::property_tree::ptree *> (sink->to_object ());
+	if (json)
+	{
+		// Format the stat data to make totals and values easier to read
+		BOOST_FOREACH (const boost::property_tree::ptree::value_type & child, json->get_child ("entries"))
+		{
+			auto time = child.second.get<std::string> ("time");
+			auto type = child.second.get<std::string> ("type");
+			auto detail = child.second.get<std::string> ("detail");
+			auto dir = child.second.get<std::string> ("dir");
+			auto value = child.second.get<std::string> ("value", "0");
+
+			if (detail == "all")
+			{
+				detail = "total";
+			}
+
+			if (type == "traffic")
+			{
+				std::stringstream numstream;
+
+				const std::vector<std::string> units = { " bytes", " KB", " MB", " GB", " TB", " PB" };
+				double bytes = std::stod (value);
+				auto index = static_cast<size_t> (std::floor (std::log2 (bytes) / 10));
+				if (index < units.size ())
```

Comment:
Shouldn't matter, but why not `std::max (index, units.size() - 1)`?

Reply:
Because the result would be a lie so I was thinking just show it raw. But yeah, don't think it matters, won't get into that range unless there's a bug.

Reply:
I think the code there will still work even if it's bigger than 1024 PB.

Reply:
Sure, I just mean it should flip over to EB at that point to be consistent. But hey, nobody will ever see it - I'll change it ;)

Reply:
Updated